tls/s2n_client_hello.c

```diff
@@ -603,3 +603,52 @@ int s2n_client_hello_get_session_id(struct s2n_client_hello *ch, uint8_t *out, u
 
     return S2N_SUCCESS;
 }
+
+static S2N_RESULT s2n_client_hello_get_raw_extension(uint16_t extension_iana,
+        struct s2n_blob *raw_extensions, struct s2n_blob *extension)
+{
+    RESULT_ENSURE_REF(raw_extensions);
+    RESULT_ENSURE_REF(extension);
+
+    struct s2n_stuffer raw_extensions_stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&raw_extensions_stuffer, raw_extensions));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&raw_extensions_stuffer, raw_extensions->size));
+
+    while (s2n_stuffer_data_available(&raw_extensions_stuffer) > 0) {
+        uint16_t extension_type = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&raw_extensions_stuffer, &extension_type));
+
+        uint16_t extension_size = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&raw_extensions_stuffer, &extension_size));
+
+        uint8_t *extension_data = s2n_stuffer_raw_read(&raw_extensions_stuffer, extension_size);
+        RESULT_ENSURE_REF(extension_data);
+
+        if (extension_iana == extension_type) {
+            RESULT_GUARD_POSIX(s2n_blob_init(extension, extension_data, extension_size));
+            return S2N_RESULT_OK;
+        }
+    }
+    RESULT_BAIL(S2N_ERR_INVALID_ARGUMENT);
+}
+
+int s2n_client_hello_has_extension(struct s2n_client_hello *ch, uint16_t extension_iana, bool *exists)
+{
+    POSIX_ENSURE_REF(ch);
+    POSIX_ENSURE_REF(exists);
+
+    *exists = false;
+
+    s2n_extension_type_id extension_type_id = s2n_unsupported_extension;
+    if (s2n_extension_supported_iana_value_to_id(extension_iana, &extension_type_id) == S2N_SUCCESS) {
+        s2n_parsed_extension *parsed_extension = NULL;
+        POSIX_GUARD(s2n_client_hello_get_parsed_extension(extension_iana, &ch->extensions, &parsed_extension));
+        *exists = true;
+        return S2N_SUCCESS;
+    }
+
+    struct s2n_blob extension = { 0 };
+    POSIX_GUARD_RESULT(s2n_client_hello_get_raw_extension(extension_iana, &ch->extensions.raw, &extension));
```

Comment:
Line 645 and line 651 mean that s2n_client_hello_has_extension will fail if 1, the extension is supported and is not found or 2, the extension is not supported and is not found. I could be reading this wrong, but doesn't this make the bool parameter redundant?

Reply:
Correct, are you suggesting `contains` API to be judged by return code, so if S2N_SUCCESS means exists, others means something wrong?

Reply:
The problem she's pointing out is that the API currently is judged by return code. Can "exists" ever be false when the s2n_client_hello_has_extension succeeds?

Reply:
Basically what we want here is to have the return code symbolize our errors, like a null pointer or other internal errors, and the boolean value stand for whether or not the extension was found. So this API shouldn't be erroring if the extension wasn't found, in that case it should succeed and the boolean value should be set to false. 

If the return code could mean either an error _or_ an extension isn't found, a user of this API could misinterpret a real error to mean that the extension wasn't found, when in fact it exists but (for example) they gave the API a null ptr.

Reply:
Oh, I know what both of you mean here. Let me update it.